Objects/unicodeobject.c

```diff
@@ -14617,6 +14617,136 @@ unicode_new_impl(PyTypeObject *type, PyObject *x, const char *encoding,
     return unicode;
 }
 
+static const char *
+as_const_char(PyObject *obj, const char *name)
+{
+    if (!PyUnicode_Check(obj)) {
+        PyErr_Format(PyExc_TypeError,
+                     "str() argument '%s' must be str, not %T",
+                     name, obj);
+        return NULL;
+    }
+    Py_ssize_t sz;
+    const char *str = PyUnicode_AsUTF8AndSize(obj, &sz);
+    if (str == NULL) {
+        return NULL;
+    }
+    if (strlen(str) != (size_t)sz) {
+        PyErr_SetString(PyExc_ValueError, "embedded null character");
+        return NULL;
+    }
+    return str;
+}
+
+static PyObject *
+fallback_to_tp_call(PyObject *type, Py_ssize_t nargs, Py_ssize_t nkwargs,
+                    PyObject *const *args, PyObject *kwnames)
+{
+    PyObject *tuple = PyTuple_New(nargs);
+    if (tuple == NULL) {
+        return NULL;
+    }
+    for (Py_ssize_t i = 0; i < nargs; i++) {
+        PyObject *value = Py_NewRef(args[i]);
+        PyTuple_SET_ITEM(tuple, i, value);
+    }
+    PyObject *dict = PyDict_New();
+    if (dict == NULL) {
+        Py_DECREF(tuple);
+        return NULL;
+    }
+    for (Py_ssize_t j = 0; j < nkwargs; j++) {
+        PyObject *key = PyTuple_GET_ITEM(kwnames, j);
+        PyObject *value = args[nargs + j];
+        if (PyDict_SetItem(dict, key, value) < 0) {
+            Py_DECREF(tuple);
+            Py_DECREF(dict);
+            return NULL;
+        }
+    }
+    return unicode_new(_PyType_CAST(type), tuple, dict);
+}
+
+static PyObject *
+unicode_vectorcall(PyObject *type, PyObject *const *args,
+                   size_t nargsf, PyObject *kwnames)
+{
+    assert(Py_Is(_PyType_CAST(type), &PyUnicode_Type));
+
+    Py_ssize_t nargs = PyVectorcall_NARGS(nargsf);
+    Py_ssize_t nkwargs = (kwnames) ? PyTuple_GET_SIZE(kwnames) : 0;
+    if (nargs == 0 && nkwargs == 0) {
+        return unicode_get_empty();
+    }
+    PyObject *object = args[0];
+    if (nargs == 1 && nkwargs == 0) {
+        return PyObject_Str(object);
+    }
+    if (nargs + nkwargs == 2) {
+        const char *encoding;
+        const char *errors;
```

Comment:
I would feel safer if both are initialized to NULL:

```suggestion
        const char *encoding = NULL;
        const char *errors = NULL;
```

Reply:
Yes, I considered this. I agree it will make the implementation easier to manage and review, so I'll change it.